SpatialGDK/Source/SpatialGDKEditor/Private/SpatialGDKEditor.cpp

```diff
@@ -50,97 +52,148 @@ bool FSpatialGDKEditor::GenerateSchema(bool bFullScan)
+		Schema::CopyWellKnownSchemaFiles(GDKSchemaCopyDir, CoreSDKSchemaCopyDir);
+		Schema::RefreshSchemaFiles(GetDefault<USpatialGDKEditorSettings>()->GetGeneratedSchemaOutputFolder());
+	}
 
-	// 80/10/10 load assets / gen schema / garbage collection.
-	FScopedSlowTask Progress(100.f, LOCTEXT("GeneratingSchema", "Generating Schema..."));
-	Progress.MakeDialog(true);
+	if (Method == CookAndGenerate)
+	{
+		// Make sure SchemaDatabase is not loaded.
+		if (UPackage* LoadedDatabase = FindPackage(nullptr, *FPaths::Combine(TEXT("/Game/"), *SpatialConstants::SCHEMA_DATABASE_FILE_PATH)))
+		{
+			TArray<UPackage*> ToUnload;
+			ToUnload.Add(LoadedDatabase);
+			UPackageTools::UnloadPackages(ToUnload);
+		}
 
-#if ENGINE_MINOR_VERSION <= 22
-	// Force spatial networking so schema layouts are correct
-	UGeneralProjectSettings* GeneralProjectSettings = GetMutableDefault<UGeneralProjectSettings>();
-	bool bCachedSpatialNetworking = GeneralProjectSettings->UsesSpatialNetworking();
-	GeneralProjectSettings->SetUsesSpatialNetworking(true);
-#endif
+		const USpatialGDKEditorSettings* EditorSettings = GetDefault<USpatialGDKEditorSettings>();
 
-	RemoveEditorAssetLoadedCallback();
+		const FString& PlatformName = EditorSettings->GetCookAndGenerateSchemaTargetPlatform();
 
-	if (Schema::IsAssetReadOnly(SpatialConstants::SCHEMA_DATABASE_FILE_PATH))
-	{
-		bSchemaGeneratorRunning = false;
-		return false;
-	}
+		if (PlatformName.IsEmpty())
+		{
+			UE_LOG(LogSpatialGDKEditor, Error, TEXT("Empty platform passed to CookAndGenerateSchema"));
+			return false;
+		}
 
-	if (!Schema::LoadGeneratorStateFromSchemaDatabase(SpatialConstants::SCHEMA_DATABASE_FILE_PATH))
-	{
-		Schema::ResetSchemaGeneratorStateAndCleanupFolders();
-	}
+		FString OptionalParams = EditorSettings->GetCookAndGenerateSchemaAdditionalArgs();
+		OptionalParams += FString::Printf(TEXT(" -targetplatform=%s"), *PlatformName);
+
+		FString ProjectPath = FPaths::IsProjectFilePathSet() ? FPaths::ConvertRelativePathToFull(FPaths::GetProjectFilePath()) : FPaths::RootDir() / FApp::GetProjectName() / FApp::GetProjectName() + TEXT(".uproject");
+		FString UATCommandLine = FString::Printf(TEXT("-ScriptsForProject=\"%s\" CookAndGenerateSchema -nocompile -nocompileeditor -server -noclient %s -nop4 -project=\"%s\" -cook -skipstage -ue4exe=\"%s\" %s -utf8output"),
+			*ProjectPath,
+			FApp::IsEngineInstalled() ? TEXT(" -installed") : TEXT(""),
+			*ProjectPath,
+			*FUnrealEdMisc::Get().GetExecutableForCommandlets(),
+			*OptionalParams
+		);
+
+		IUATHelperModule::Get().CreateUatTask(UATCommandLine,
+			FText::FromString(PlatformName),
+			LOCTEXT("CookAndGenerateSchemaTaskName", "Cook and generate project schema"),
+			LOCTEXT("CookAndGenerateSchemaTaskName", "Generating Schema"),
+			FEditorStyle::GetBrush(TEXT("MainFrame.PackageProject")),
+			[](FString ResultType, double RuntimeInSec)
+			{
+				if (ResultType == TEXT("Completed"))
+				{
+					//??
```

Comment:
?

Reply:
Forgot that, I was wondering what I should do In case of success. Maybe notify the asset manager about the new schema database asset at least.

Reply:
If you need to do that, then probably a good idea. otherwise can just remove the branch